python/GafferSceneUI/SceneViewUI.py

```diff
@@ -1073,13 +1076,105 @@ def __appendClippingPlaneMenuItems( menuDefinition, prefix, view, parentWidget )
 			}
 		)
 
+def __snapshotDescription( view ) :
+
+	sceneGadget = view.viewportGadget().getPrimaryChild()
+	if sceneGadget.getRenderer() == "OpenGL" :
+		return "Viewport snapshots are only available for rendered (non-OpenGL) previews."
+
+	return "Snapshot viewport and send to catalogue."
+
+def __snapshotToCatalogue( catalogue, view ) :
+
+	timeStamp = str( datetime.datetime.now() )
+
+	fileName = pathlib.Path(
+		view["in"].node().getContext().substitute( catalogue["directory"].getValue() )
+	) / ( timeStamp.replace( ":", "-" ).replace(" ", "_" ) + ".exr" )
+
+	resolutionGate = imath.Box3f()
+	if isinstance( view, GafferSceneUI.SceneView ) :
+		resolutionGate = view.resolutionGate()
+
+	scriptRoot = view["in"].getInput().ancestor( Gaffer.ScriptNode )
+	metadata = IECore.CompoundData(
+		{
+			"gaffer:sourceScene" : view["in"].getInput().relativeName( scriptRoot ),
+			"gaffer:context:frame" : view["in"].node().getContext().getFrame()
+		}
+	)
+
+	sceneGadget = view.viewportGadget().getPrimaryChild()
+	sceneGadget.snapshotToFile( fileName, resolutionGate, metadata )
+
+	image = GafferImage.Catalogue.Image( "Snapshot1", Gaffer.Plug.Direction.In, Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic )
+	image["fileName"].setValue( fileName )
+
+	image["description"].setValue(
+		"Snapshot of {} at frame {} taken at {}".format(
+			metadata["gaffer:sourceScene"],
+			metadata["gaffer:context:frame"], timeStamp[:-6]  # Remove trailing microseconds
+		)
+	)
+
+	catalogue["images"].source().addChild( image )
+
+def __snapshotCataloguesSubMenu( view, scriptNode ) :
+
+	menuDefinition = IECore.MenuDefinition()
+
+	catalogueList = list( GafferImage.Catalogue.RecursiveRange( scriptNode ) )
+
+	if len( catalogueList ) == 0 :
+		menuDefinition.append(
+			"/No Catalogues Available",
+			{
+				"active" : False,
+			}
+		)
+
+	else :
+		snapshotDescription = __snapshotDescription( view )
+		snapshotActive = view["renderer"]["name"].getValue() != "OpenGL"
+
+		for c in catalogueList :
+
+			cName = c["name"].getValue()
+			nName = c.relativeName( scriptNode )
+
+			menuDefinition.append(
+				"/" + ( nName + ( " ({})".format( cName ) if cName else "" ) ),
+				{
+					"active" : snapshotActive,
```

Comment:
`and not Gaffer.MetadataAlgo.readOnly( c )`.

Reply:
I think it should be even more specific to `c["imageIndex"].source()` and `c["images"].source()`? I'm doing that in https://github.com/GafferHQ/gaffer/pull/5533/commits/3d73663775ffcc85463d4b23c8b8f35229586c57

I think if we just look at `c` then any time the catalog is in a Reference it won't be editable, correct?

Reply:
Yep, spot on. Thanks once again for translating my gibberish into something mergeable.